README.md
# Carve a jack-o-lantern in R

Happy Halloween 2023!

I made this package partly for fun (who doesn't want to draw a jack-o-lantern in R?) and partly as a simple example of how to write a package. I added the face image to my drawing last year (2022) and cleaned it up this year (2023) when I converted my drawing to a package. I also added a cat this year, so your pumpkins can have some variety. Maybe I'll add something else next year.

To use the package, install it, then run the only function that is exported, `draw_jackolantern()`.

Warning: it's graphic-heavy and not optimized, so it may take several seconds to run. I suggest saving the image first with `p <- draw_jackolantern()`, which should run quickly, then running `print(p)` separately, since the drawing part is what takes a little while and occasionally, only half the image will be drawn. Eventually, I may fix that.

Feel free to download and use this package. Let me know how it works for you (or doesn't) by posting an issue. 



NEWS.md
# carvedpumpkin v0.1.0

* Converted 2022 image to a package
* Added cat image as carving option



R/create_face.R
#' Carve a smiling face
#'
#' @param image A ggplot object.
#'
#'

create_face <- function(image) {
  # eyes - mirror ----
  eye_left <- data.frame(
    x = c(0, 40, 50, 65, 90, 100, 120, 140, 120, 100, 90, 50, 20) + 290,
    y = c(0, 5, 30, 40, 40, 35, 5, 0, 80, 100, 105, 95, 70) + 310
  )
  eye_right <- eye_left
  eye_right$x = 1000 - eye_right$x

  # mouth - mirrors itself ----
  mouth_left <- data.frame(
    x = c(500, 450, 430, 410, 390, 350, 320, 300, 320, 340, 350, 360, 390,
          420, 430, 460, 490, 500),
    y = c(180, 180, 190, 220, 200, 220, 230, 250, 250, 240, 250, 250, 230,
          250, 250, 220, 250, 250),
    id = 1:18
  )
  mouth_right <- mouth_left
  mouth_right$x = 1000 - mouth_right$x
  mouth_right <- mouth_right[order(mouth_right$id, decreasing = TRUE), ]
  mouth_right <- mouth_right[!mouth_right$x == 500, ]
  mouth <- rbind(mouth_left, mouth_right)
  rm(mouth_left, mouth_right)

  # candle ----
  candle <- data.frame(
    x = c(490, 500, 510, 500, 490, 480),
    y = c(180, 180, 190, 230, 240, 200) + 5
  )

  # add face to image ----
  p <- image + # pumpkin: eyes
    ggforce::geom_shape(eye_left,
                        mapping = ggplot2::aes(x = !!dplyr::sym("x"),
                                               y = !!dplyr::sym("y")),
                        radius = grid::unit(0.012, "native"),
                        color = "darkorange", fill = "yellow") +
    ggforce::geom_shape(eye_right,
                        mapping = ggplot2::aes(x = !!dplyr::sym("x"),
                                               y = !!dplyr::sym("y")),
                        radius = grid::unit(0.012, "native"),
                        color = "darkorange", fill = "yellow") +
    # pumpkin: mouth
    ggforce::geom_shape(mouth,
                        mapping = ggplot2::aes(x = !!dplyr::sym("x"),
                                               y = !!dplyr::sym("y")),
                        radius = grid::unit(0.003, "native"),
                        color = "darkorange", fill = "yellow")

  # add candle to image ----
  p <- p + # pumpkin: candle
    ggforce::geom_shape(candle,
                        mapping = ggplot2::aes(x = !!dplyr::sym("x"),
                                               y = !!dplyr::sym("y")),
                        radius = grid::unit(0.01, "native"), fill = "gold1",
                        color = "yellow", expand = grid::unit(0.5, 'mm')) +
    ggforce::geom_shape(candle,
                        mapping = ggplot2::aes(x = !!dplyr::sym("x"),
                                               y = !!dplyr::sym("y")),
                        radius = grid::unit(0.005, "native"), fill = "darkgoldenrod1",
                        color = "gold1", expand = grid::unit(-0.5, 'mm')) +
    ggforce::geom_shape(candle,
                        mapping = ggplot2::aes(x = !!dplyr::sym("x"),
                                               y = !!dplyr::sym("y")),
                        radius = grid::unit(0, "native"), fill = "darkorange",
                        color = "darkgoldenrod1", expand = grid::unit(-1, 'mm'))

  # return image ----
  return(p)
}


R/create_pumpkin.R
#' Grow the pumpkin
#'
#' @description
#' Sorry, all pumpkins are cloned from the same stock, so there is no variation.
#'

create_pumpkin <- function() {
  # backdrop ----
  sky_gradient <- grDevices::colorRampPalette(c("darkgreen", "blue4",
                                                "purple4"))(100)
  bg <- tibble::tibble(x = (0:100) * 10, y = (0:100) * 7)
  stars <- tibble::tibble(x = sample(seq(0, 1000), 500, replace = TRUE),
                          y = sample(seq(250, 700), 500, replace = TRUE),
                          vals = factor(sample(1:10, 500, replace = TRUE)))
  stars$x100 <- floor(stars$x / 100)
  stars$y100 <- floor(stars$y / 100)

  # stem ----
  stem <- data.frame(
    id = rep(1:3, each = 4),
    x = c(470, 480, 490, 485,  485, 490, 505, 515,  515, 505, 515, 530),
    y = c(480, 520, 524, 480,  480, 524, 526, 480,  480, 526, 530, 480),
    vals = factor(rep(c(1,2,1), each = 4)), type = 2
  )

  # plot object ----
  graphics::par(bg = "black", mar = c(0, 0, 0, 0), mai = c(0, 0, 0, 0))
  # background ----
  p <- ggplot2::ggplot(stars, ggplot2::aes(x = !!dplyr::sym("x"),
                                           y = !!dplyr::sym("y"))) +
    ggplot2::theme_classic() +
    ggplot2::geom_line() +
    mapply(function(ylow, yhigh, col, a = 1) {
      ggplot2::geom_ribbon(ggplot2::aes(ymin = ylow, ymax = yhigh),
                           alpha = a, fill = col)
    }, utils::head(bg$y, -1), utils::tail(bg$y, -1), sky_gradient,
    seq(1, 700, length = 100)) +
    ggplot2::theme(axis.text = ggplot2::element_blank(),
                   axis.ticks = ggplot2::element_blank(),
                   axis.title = ggplot2::element_blank(),
                   axis.line = ggplot2::element_blank(),
                   legend.position = "none")
  p <- p + # stars
    ggplot2::geom_point(size = 1, colour = "white", shape = ".", alpha = 1)
  # pumpkin ----
  p <- p + # pumpkin: stem
    ggplot2::geom_polygon(stem, mapping =
                            ggplot2::aes(x = !!dplyr::sym("x"),
                                         y = !!dplyr::sym("y"),
                                         fill = !!dplyr::sym("vals"),
                                         group = !!dplyr::sym("vals"))) +
    ggplot2::scale_fill_manual(values = c("chartreuse4", "mediumseagreen")) +
    # pumpkin body: outside edges
    ggforce::geom_ellipse(color = "chocolate4", fill = "darkorange3",
                          ggplot2::aes(x0 = 350, y0 = 300, a = 250, b = 200,
                                       angle = 0)) +
    ggforce::geom_ellipse(color = "chocolate4", fill = "darkorange3",
                          ggplot2::aes(x0 = 650, y0 = 300, a = 250, b = 200,
                                       angle = 0)) +
    # pumpkin body: next layer
    ggforce::geom_ellipse(color = "chocolate3", fill = "darkorange2",
                          ggplot2::aes(x0 = 400, y0 = 300, a = 180, b = 200,
                                       angle = 0)) +
    ggforce::geom_ellipse(color = "chocolate3", fill = "darkorange2",
                          ggplot2::aes(x0 = 600, y0 = 300, a = 180, b = 200,
                                       angle = 0)) +
    # pumpkin body: middle layer
    ggforce::geom_ellipse(color = "chocolate2", fill = "darkorange",
                          ggplot2::aes(x0 = 500, y0 = 300, a = 150, b = 200,
                                       angle = 0)) +
    ggplot2::coord_fixed()
  # return image ----
  return(p)
}


R/create_cat.R
#' Carve a cat
#'
#'
#' @param image A ggplot object.

create_cat <- function(image) {
  # cat - on broom? or with hat?
  xplus <- 320
  yplus <- 160
  # legs ----
  leftfore <- data.frame(
    x = c( 0, 20, 30, 40, 50, 40, 10,  5) * 3 + xplus,
    y = c(20, 30, 30, 40, 30, 20, 20, 15) * 3 + yplus
  )
  rightfore <- data.frame(
    x = c(15, 30, 30, 40, 50, 40, 25) * 3 + xplus,
    y = c( 0, 20, 30, 40, 30, 20,  0) * 3 + yplus
  )
  hind <- data.frame(
    x = c(35, 60, 60, 70, 80, 80, 70, 65, 35) * 3 + xplus,
    y = c( 5, 10, 20, 25, 20, 10, 10,  0,  0) * 3 + yplus
  )
  # tail ----
  tail <- data.frame(
    x = c(90, 90, 85, 90, 100, 100, 90, 100, 100, 95) * 3 + xplus,
    y = c(20, 40, 60, 80,  80,  70, 65,  40,  30, 20) * 3 + yplus
  )
  # head ----
  head <- data.frame(
    x = c(15, 20, 25, 30, 35, 35, 45, 50, 50, 20) * 3 + xplus,
    y = c(60, 60, 70, 75, 75, 85, 80, 70, 50, 50) * 3 + yplus
  )
  # body ----
  body <- data.frame(
    x = c(30, 50, 80, 90, 95, 90, 70, 30) * 3 + xplus,
    y = c(50, 55, 40, 30, 20, 10, 10, 30) * 3 + yplus
  )
  # candle ----
  candle <- data.frame(
    x = c(50, 45, 45, 50, 50, 55, 50, 55) * 3 + xplus,
    y = c(20, 25, 30, 35, 40, 35, 30, 25) * 3 + yplus
  )

  # draw image ----
  p <- image + # pumpkin: cat body
    ggforce::geom_shape(head,
                        mapping = ggplot2::aes(x = !!dplyr::sym("x"),
                                               y = !!dplyr::sym("y")),
                        radius = grid::unit(0.003, "native"),
                        color = "yellow", fill = "yellow") +
    ggforce::geom_shape(tail,
                        mapping = ggplot2::aes(x = !!dplyr::sym("x"),
                                               y = !!dplyr::sym("y")),
                        radius = grid::unit(0.005, "native"),
                        color = "yellow", fill = "yellow") +
    ggforce::geom_shape(body,
                        mapping = ggplot2::aes(x = !!dplyr::sym("x"),
                                               y = !!dplyr::sym("y")),
                        radius = grid::unit(0.02, "native"),
                        color = "yellow", fill = "yellow") +
    # pumpkin: cat legs
    ggforce::geom_shape(leftfore,
                        mapping = ggplot2::aes(x = !!dplyr::sym("x"),
                                               y = !!dplyr::sym("y")),
                        radius = grid::unit(0.008, "native"),
                        color = "yellow", fill = "yellow") +
    ggforce::geom_shape(rightfore,
                        mapping = ggplot2::aes(x = !!dplyr::sym("x"),
                                               y = !!dplyr::sym("y")),
                        radius = grid::unit(0.01, "native"),
                        color = "yellow", fill = "yellow") +
    ggforce::geom_shape(hind,
                        mapping = ggplot2::aes(x = !!dplyr::sym("x"),
                                               y = !!dplyr::sym("y")),
                        radius = grid::unit(0.01, "native"),
                        color = "yellow", fill = "yellow")

  # add candle to image ----
  p <- p + # pumpkin: candle
    ggforce::geom_shape(candle,
                        mapping = ggplot2::aes(x = !!dplyr::sym("x"),
                                               y = !!dplyr::sym("y")),
                        radius = grid::unit(0.01, "native"), fill = "gold1",
                        color = "yellow", expand = grid::unit(0.5, 'mm')) +
    ggforce::geom_shape(candle,
                        mapping = ggplot2::aes(x = !!dplyr::sym("x"),
                                               y = !!dplyr::sym("y")),
                        radius = grid::unit(0.005, "native"), fill = "darkgoldenrod1",
                        color = "gold1", expand = grid::unit(-0.5, 'mm')) +
    ggforce::geom_shape(candle,
                        mapping = ggplot2::aes(x = !!dplyr::sym("x"),
                                               y = !!dplyr::sym("y")),
                        radius = grid::unit(0, "native"), fill = "darkorange",
                        color = "darkgoldenrod1", expand = grid::unit(-1, 'mm'))

  # return image ----
  return(p)
}
